background-processor/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+/// * Calling `ChannelManager::timer_chan_freshness_every_min()` every minute (can be done in the
+///   background).
+///
+/// Note that if ChannelManager persistence fails and the persisted manager becomes out-of-date,
+/// then there is a risk of channels force-closing on startup when the manager realizes it's
+/// outdated. However, as long as `ChannelMonitor` backups are sound, no funds besides those used
+/// for unilateral chain closure fees are at risk.
+///
+/// [`PersistManager`]: type.PersistManager.html
+pub struct BackgroundProcessor {
+	stop_thread: Arc<AtomicBool>,
+	/// May be used to retrieve and handle the error if `BackgroundProcessor`'s thread
+	/// panics.
+	pub thread_handle: JoinHandle<()>
+}
+
+// Re-export these modules for the logging macros.
+mod util {
+	pub(crate) mod logger {
+		pub(crate) use lightning::util::logger::Record;
+	}
+}
+
+#[cfg(not(test))]
+const CHAN_FRESHNESS_TIMER: u64 = 60;
+#[cfg(test)]
+const CHAN_FRESHNESS_TIMER: u64 = 1;
+
+/// Persist a `ChannelManager`: this could mean writing once to disk, and/or uploading to one or
+/// more backup services. This function is passed to the `BackgroundProcessor` on startup. See
+/// [`ChannelManager::write`] for writing out a `ChannelManager`.
+///
+/// [`ChannelManager::write`]: ../lightning/ln/channelmanager/struct.ChannelManager.html#method.write
+pub type PersistManager<ChanSigner, M, T, K, F, L> = fn(String, Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>;
+
+impl BackgroundProcessor {
+	/// Start the background thread that takes care of responsibilities (enumerated in the top-level
+	/// documentation). Marked as `must_use` because otherwise the result is dropped immediately,
+	/// resulting in the thread being terminated.
+	/// Important note: this thread will panic if invoking `persist_manager` results in an error (and
+	/// `start()` will need to be called again to restart the `BackgroundProcessor`).
+	/// There are 3 main options for handling this panic:
+	/// * wait on [`thread_handle`]'s `join()`, handle the error
+	/// * [configure] to abort on panic
+	/// * write a custom [`PersistManager`] to handle the error so it never gets returned to
+	///   `BackgroundProcessor`.
+	///
+	/// [`thread_handle`]: struct.BackgroundProcessor.html#structfield.thread_handle
+	/// [configure]: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html
+	/// [`PersistManager`]: type.PersistManager.html
+	#[must_use]
+	pub fn start<ChanSigner: 'static + ChannelKeys + Writeable, M: 'static + chain::Watch<Keys=ChanSigner>, T: 'static + BroadcasterInterface, K: 'static + KeysInterface<ChanKeySigner=ChanSigner>, F: 'static + FeeEstimator, L: 'static + Logger>(data_dir_path: String, persist_manager: PersistManager<ChanSigner, M, T, K, F, L>, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>, logger: Arc<L>) -> Self {
```

Comment:
`data_dir_path` seems specific to `FilesystemPersister`. Should a caller need to supply this if their persister doesn't need it?  Instead of including this parameter, you could probably use a closure at the call site if there are issues with needing a `self` parameter.

https://stackoverflow.com/questions/24728394/is-there-a-way-to-create-a-function-pointer-to-a-method-in-rust

Reply:
Nice article. I can use the solution in that link to get access to the `self` parameter. Unfortunately then we run into the issue that the closure may outlive the `self` variable that it captures, since the closure is used in the bg thread so its lifetime has to be `'static` (and I don't think there's a way to ensure the captured variables are 'static? Am I missing something here?). 

Let me know what you think.

Reply:
> Nice article. I can use the solution in that link to get access to the `self` parameter. Unfortunately then we run into the issue that the closure may outlive the `self` variable that it captures, since the closure is used in the bg thread so its lifetime has to be `'static` (and I don't think there's a way to ensure the captured variables are 'static? Am I missing something here?).
> 
> Let me know what you think.
> 

Instead of capturing `self`, you can make a local variable containing the result of `get_data_dir` and capture that, I believe.

> I agree the `data_dir_path` is awkward though. I sort of prefer the original `trait` solution.

Likewise, as I think it is cleaner than using the lambda approach. :)

Reply:
Doh. Sounds good, I followed your last comment.